### DISCUSSION TITLE
ipxe.efi boot from USB stuck in initializing devices, want to boot Windows 10 from iSCSI in EFI on non-EFI nic

### DISCUSSION THREAD

**Discussion by janisbit** [ipxe/ipxe#1059] [General]
Hello,
I have a ConnectX-3 NIC  (specifically a  HP branded MCX354A-FCB_A2-A5) which unfortunately lacks an UEFI boot ROM, and it seems hard / impossible to find one since the cards are EOL and by default  it does not ship with a UEFI boot ROM.

See my recent related question on servethehome.com
https://forums.servethehome.com/index.php?threads/connectx-3-uefi-rom.33131/#post-396214

I would like to ask if it possible to do the following:  
put ipxe.efi on a USB stick, boot the PC in EFI mode from and start ipxe.efi and from there configure the Connect-X3 and load a iSCSI-aware WinPE in order to install Windows on the iSCSI target in EFI mode since it supports disk greater than 2TB.
I already succeded booting Windows 10 from a iSCSI target in BIOS mode but with the limitation of 2TB.

So is the USB Stick boot in EFI mode and sanhook via the ConnectX-3 for installing / booting Windows 10 in EFI mode technically possible or is there a limitation in ipxe which makes this impossible?

on my Mainboard ASUS B450M H - Ryzen CPU,  when I create an EFI partition on an USB stick I must rename the ipxe.efi to
bootx64.efi  otherwise it does not start. It must reside in \EFI\BOOT

Unfortunately when starting it remains stuck in iPXE initializing devices ...
(I compiled it from git a few days ago)

(The mainboard has a integrated Realtek 1GbE NIC which I don't use.)

So I recompiled ipxe.efi with DEBUG options to see what's going on.

make bin-x86_64-efi/ipxe.efi  DEBUG="device,runtime,pci,init,nii,snp,snpnet,snponly,efi_watchdog,efi_image,image_cmd,image,efi_autoboot,efi_timer,efi_driver:3,efi_debug,efi_init,efi_pci,efi_wrap,intel" LOG_LEVEL=LOG_ALL

Unfortunately the debug messages list is longer than one screen and the messages do scroll fast, so I made a short video and extracted a few frames. 

https://ibb.co/xSx1JFw
https://ibb.co/VS4XCX8
https://ibb.co/mhj6Z35
https://ibb.co/1KJx8Y2

Should I proceed further with debugging or is my use case not covered by ipxe?

One more thing I tried is to create an USB stick with and iSCSI enabled WinPE using the Windows 10 ADK and it boots and by quering the registry I see it booted in UEFI mode. I can mount the iSCSI target using iscsicpl but Windows 10 setup.exe refuses to install on that disk because it says " "Windows cannot be installed to this disk. iSCSI deployment is disabled since no NICs referenced in the iBFT can be resolved to actual NT-visible devices."

many thanks in advance for your help.
Best regards,
Janis

**Comment by NiKiZe** [comment]
Yes you need to rename it to bootx64.efi
Use less debug options 
Your images has ghosting and are mostly unreadable.
You can upload images directly to GitHub.

Try building the .efi for your specific nic, give your pcivendor and pcidevice IDs as a 8char string. Example at 
https://ipxe.org/appnote/buildtargets#driver

I do suspect that your Realtek is involved in the hang.
Test snponly.efi and then snp.efi to see if they hang, your wanted NIC won't work but it will give more data on if you can get around the hang.

**Comment by janisbit** [comment]
Hello Christian,
many thanks for your great answers.
I made some additional tests as you advised:
I compiled snp.efi and snponly.efi  and put it no an USB stick and tried to boot with a network cable plugged into the Realtek (on board) NIC.

snponly.efi:
it boots but it says no network devices and then I see the ipxe prompt where you can enter commands.

snp.efi:
it boots and then sees the realtek as net0,  gets a DHCP address from the DHCP server.  For the realtek macaddr I set an entry filename "ipxe.efi"  on the DHCP server.   it executes it but then it hangs (like loading ipxe.efi natively from the USB stick).

So it seems like the Realtek is not the problem?


For example in the Mainboards BIOS I have options:
LAN PXE Boot Option ROM:  I set it to disabled since I assume it is related to the Realtek nic boot
Storage Boot Option Control: Uefi Only  (to boot from USB stick in UEFI mode. It has options Uefi Only, Legacy Only, Disabled).
Other PCI Device ROM Priority: Disabled  ( It has options Disabled, UEFI Only, Legacy Only).
Do you think the "Other PCI Device ROM" setting controls the ROM priority of the Mellanox card too ?
Anyway, since my Mellanox card does not have a UEFI ROM, I think this setting does not matter since in this case we boot from USB.

The BIOS has a Network Stack enable/disable option.  I set it to disable to see if something changes but ipxe.efi still hangs.

As advised, as a next step I compiled ipxe with only  the Mellanox ConnectX-3 driver  15b31003.efi
with more and less debug options, ran it from the USB stick. it unfortunately still hangs in initializing devices.

1)
make bin-x86_64-efi/15b31003.efi  DEBUG="device,runtime,init,nii,efi_watchdog,efi_image,image_cmd,image,efi_autoboot,efi_timer,efi_driver:3,efi_debug,efi_init,efi_pci" LOG_LEVEL=LOG_ALL

Screen output:
https://ibb.co/tCDgVyV

2)
make bin-x86_64-efi/15b31003.efi  DEBUG="device,runtime,pci,init,nii,snp,snpnet,snponly,efi_watchdog,efi_image,image_cmd,image,efi_autoboot,efi_timer,efi_driver:3,efi_debug,efi_init,efi_pci" LOG_LEVEL=LOG_ALL

https://ibb.co/gD2BTyX


I tried all the above Mainboard BIOS Settings combinations but no difference.
Unfortunately I cannot make a clear video of the full debug output because of monitor ghosting.
If it is important for the resolution of the problem, is there a way to send the debug output to a serial port? Would an USB-to-serial converter work?

Note that this Mellanox card has still the original HP firmware (as I posted on servethehome.com). I have another identical card where I reflashed the standard Mellanox firmware. I could try to test if it makes a difference.
But on this card with original HP firmware,  WinPE works correctly form USB. Boots from USB in EFI mode and brings up the network. ping and other network functions do work.

Any additional things I should try in order to investigate the cause of the hang? 

Many thanks in advance.
Best regards,
Janis

**Reply by NiKiZe** [reply]
With the SNP tests we now know that some driver in iPXE, probably the Realtek one, causes hang.

15b31003.efi hanging is concerning. 

Stop adding so much debug options.

Does 15b31003.efi hang without any debug options? And does it still hang with network stack disabled in efi firmware?

**Comment by janisbit** [comment]
Hello,
tried to boot 15b31003.efi  without debug with Mainboard BIOS  Network Stack Enabled/Disabled,   PXE Boot ROM Enabled/Disabled, all 4 combinations.
In each case
it show only iPXE initialising devices...  and hangs.

What do you mean with " network stack disabled in efi firmware" ? Disable something while compiling ipxe or did you mean the Mainboard BIOS Option?

thanks,
Janis

**Comment by janisbit** [comment]
I made some further tests:
I bootet the USB stick in EFI mode with 15b31003.efi and it boots normally and I get to the ipxe prompt.

What I should not is that since the Ryzen PC with the ASUS B450M H has only 1 PCI-e 3.0 x16 slot and 2 PCI-e 2.0 x1 slots and the x16 slot is occupied by the GPU,  I used a M.2 to PCI-e adapter (rated PCI-e 3.0 x4)  to connect the ConnectX-3 in x4 mode.

This is the adapter. 
https://www.adt.link/product/R43M.html

I did some performance tests in Windows and Linux and did look at the output of dmesg in Linux and 
the card's bandwidth is downgraded to 2.5GT/s   x4,  which means PCI-e1 and  iperf gives me a peak of about 900MB/s, about 7Gbit/s which is compatible with the selected PCI speed.

What's interesting is that when using a script to set the PCI speed (version 1,2 or 3):
https://www.alexforencich.com/wiki/en/pcie/set-speed

If I set the speed to PCI-e 3 then it tries to reconfigure the speed but fails and switches back to PCI-e 1.
If I set the speed to PCI-e 2 then it succeds and indeed the iperf bandwidth doubles ( about 13-14Gbit).

If I connect a nvme drive using a m.2 to PCI-e riser to this adapter then it works at full PCI-e 3 speed.

But in case of the ConnectX-3 something strange is going on because it drops the PCI-e speed.

Perhaps this is the cause why ipxe locks up in EFI mode?   
But ipxe works in BIOS mode. 
As said, I used the integrated Flexboot to boot a "pxelinux.0" with a menu and from there load ipxe.lkrn and everything works fine (except for the 2TB iSCSI disk limit in MBR mode which I try to solve by using EFI).

WinPE,  Windows 10,  Linux do not have a problem with the ConnectX-3 card connected to the M.2 to PCI-e adapter, except to the speed reduced to PCI-e 1 (or PCI-e 2 with the set_pci_speed.sh).

Is it possible to compile the ipxe driver 15b31003.efi  and put it directly on the ConnectX-3 card as an expansion ROM or is this not supported?  

Alternatively  I could try this:

since it seems hard to find a expansion ROM for Flexboot with UEFI support for the MCX354A-FCB_A2-A5, would it possible to compile it from source?

Is this the correct repository ?
[GitHub - ipxe/vendor-mellanox-flexboot: Mellanox FlexBoot public repository](https://github.com/ipxe/vendor-mellanox-flexboot)

It seems updated last 7 years, it says Flexboot 3.4.812, so the major.minor version (3.4) is the same that is on my card, which in theory should work.
By using a Linux Debian 10, I managed to compile the ROM for the ConnectX-3 (see the README in the src directory how to build the ROM).

But this ROM file is not signed so I cannot burn it to the card.
Any idea how to sign it? I tried to search how do do it but I did not find the relevant information yet.

Any hints how to proceed?

Many thanks in advance,
Janis
